pkg/controllers/kepler_internal.go

```diff
@@ -0,0 +1,403 @@
+
+// RBAC for running Kepler exporter
+//+kubebuilder:rbac:groups=apps,resources=daemonsets,verbs=list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=security.openshift.io,resources=securitycontextconstraints,verbs=list;watch;create;update;patch;delete;use
+//+kubebuilder:rbac:groups=monitoring.coreos.com,resources=servicemonitors;prometheusrules,verbs=list;watch;create;update;patch;delete
+
+// RBAC required by Kepler exporter
+//+kubebuilder:rbac:groups=core,resources=nodes/metrics;nodes/proxy;nodes/stats,verbs=get;list;watch
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KeplerInternalReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	// We only want to trigger a reconciliation when the generation
+	// of a child changes. Until we need to update our the status for our own objects,
+	// we can save CPU cycles by avoiding reconciliations triggered by
+	// child status changes.
+	//
+	// TODO: consider using ResourceVersionChanged predicate for resources that support it
+
+	genChanged := builder.WithPredicates(predicate.GenerationChangedPredicate{})
+
+	c := ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.KeplerInternal{}).
+		Owns(&corev1.ConfigMap{}, genChanged).
+		Owns(&corev1.ServiceAccount{}, genChanged).
+		Owns(&corev1.Service{}, genChanged).
+		Owns(&appsv1.DaemonSet{}, builder.WithPredicates(predicate.ResourceVersionChangedPredicate{})).
+		Owns(&rbacv1.ClusterRoleBinding{}, genChanged).
+		Owns(&rbacv1.ClusterRole{}, genChanged)
+
+	if r.Cluster == k8s.OpenShift {
+		c = c.Owns(&secv1.SecurityContextConstraints{}, genChanged)
+	}
+	return c.Complete(r)
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Kepler object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *KeplerInternalReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// TODO: remove these keys from the log
+	// "controller": "kepler", "controllerGroup": "kepler.system.sustainable.computing.io",
+	// "controllerKind": "Kepler", "Kepler": {"name":"kepler"},
+
+	logger := log.FromContext(ctx)
+	r.logger = logger
+
+	logger.Info("Start of  reconcile")
+	defer logger.Info("End of reconcile")
+
+	kepler, err := r.getInternal(ctx, req)
+	if err != nil {
+		// retry since some error has occurred
+		logger.V(6).Info("Get Error ", "error", err)
+		return ctrl.Result{}, err
+	}
+
+	if kepler == nil {
+		// no kepler found , so stop here
+		logger.V(6).Info("Kepler Nil")
+		return ctrl.Result{}, nil
+	}
+
+	logger.V(6).Info("Running sub reconcilers", "kepler", kepler.Spec)
```

Comment:
running kepler internal ?

Reply:
same comment for other logs in this file, should it say `kepler` or `keplerinternal` ?